api/workload/v2/reader.go

```diff
@@ -0,0 +1,83 @@
+package workload
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/proto/api/workload"
+)
+
+type streamReader struct {
+	SVIDChan      chan *workload.X509SVIDResponse
+	logger        *logrus.Logger
+	streamManager *streamManager
+	ctx           context.Context
+	cancelFn      func()
+
+	// atomic stream
+	stream *managedStream
+}
+
+func newStreamReader(ctx context.Context, logger *logrus.Logger, streamManager *streamManager) *streamReader {
+	ctx, cancel := context.WithCancel(ctx)
+	r := &streamReader{
+		ctx:           ctx,
+		cancelFn:      cancel,
+		logger:        logger,
+		streamManager: streamManager,
+		SVIDChan:      make(chan *workload.X509SVIDResponse),
+	}
+	r.start()
+	return r
+}
+
+func (c *streamReader) start() {
+	c.logger.Debug("Starting reader.")
+	go func() {
+		defer c.logger.Debug("Shutting down reader")
+		defer c.closeStream()
+		defer close(c.SVIDChan)
+
+		for {
+			select {
+			case stream, ok := <-c.streamManager.StreamChan:
+				if !ok {
+					return
+				}
+				c.stream = stream
+			case <-c.ctx.Done():
+				return
+			}
+
+			for {
+				resp, err := c.stream.Recv(c.ctx)
+				if err != nil {
+					c.logger.WithError(err).Info("Stream reader failed.")
```

Comment:
This should likely check for `io.EOF` and log something different in the known shutdown path.

Reply:
Hmmm then you would also want to check if the context was cancelled. This is fine as is.